File: src/pts-repair.c
/*
 * obs-irl-source — IRL streaming source plugin for OBS
 * https://irlserver.com
 *
 * Copyright (C) 2026 Thomas Lekanger
 * SPDX-License-Identifier: AGPL-3.0-or-later
 *
 * pts-repair.c — PTS discontinuity detection and repair
 *
 * Three gap ranges:
 *   small  (<  small_gap_ms):  interpolate from previous PTS
 *   medium (>= small_gap_ms, < large_gap_ms): insert silence
 *   large  (>= large_gap_ms): full timestamp reset
 */

#include "../include/pts-repair.h"

/* ── Helpers ──────────────────────────────────────────────── */

static int ts_to_ms(const struct pts_repair *r, int64_t ts)
{
	if (r->tb_den == 0)
		return 0;
	return (int)(ts * 1000 * r->tb_num / r->tb_den);
}

/* ── Public API ───────────────────────────────────────────── */

void pts_repair_init(struct pts_repair *r, int small_gap_ms, int large_gap_ms,
		     int tb_num, int tb_den)
{
	r->last_pts = 0;
	r->last_duration = 0;
	r->tb_num = tb_num;
	r->tb_den = tb_den;
	r->small_gap_ms = small_gap_ms;
	r->large_gap_ms = large_gap_ms;
	r->initialised = false;
}

void pts_repair_reset(struct pts_repair *r)
{
	r->last_pts = 0;
	r->last_duration = 0;
	r->initialised = false;
}

enum pts_action pts_repair_evaluate(struct pts_repair *r, int64_t pts,
				    int64_t duration,
				    int64_t *corrected_pts,
				    int *silence_ms)
{
	*silence_ms = 0;

	/* First frame — just record and pass through */
	if (!r->initialised) {
		r->last_pts = pts;
		r->last_duration = duration > 0 ? duration : 1;
		r->initialised = true;
		*corrected_pts = pts;
		return PTS_ACTION_PASS;
	}

	/* Expected PTS = last_pts + last_duration */
	int64_t expected = r->last_pts + r->last_duration;
	int64_t gap = pts - expected;

	/* Convert gap to milliseconds for threshold comparison */
	int gap_ms = ts_to_ms(r, gap >= 0 ? gap : -gap);
	bool is_backward = gap < 0;

	/* Backward jump or tiny gap — likely reorder, pass through */
	if (is_backward || gap_ms < 1) {
		r->last_pts = pts;
		r->last_duration = duration > 0 ? duration : r->last_duration;
		*corrected_pts = pts;
		return PTS_ACTION_PASS;
	}

	enum pts_action action;

	if (gap_ms < r->small_gap_ms) {
		/* Small gap — interpolate: use expected PTS */
		*corrected_pts = expected;
		action = PTS_ACTION_INTERPOLATE;
	} else if (gap_ms < r->large_gap_ms) {
		/* Medium gap — insert silence, then use original PTS */
		*corrected_pts = pts;
		*silence_ms = gap_ms;
		action = PTS_ACTION_SILENCE;
	} else {
		/* Large gap — full reset */
		*corrected_pts = pts;
		action = PTS_ACTION_RESET;
	}

	r->last_pts = *corrected_pts;
	r->last_duration = duration > 0 ? duration : r->last_duration;

	return action;
}

File: crates/irl-core/src/pts_repair.rs
//! Three-tier PTS discontinuity repair (port of `src/pts-repair.c`).

/// What the caller should do with a frame.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PtsAction {
    /// Use the PTS as is.
    Pass,
    /// PTS was repaired by interpolation.
    Interpolate,
    /// Insert `silence_ms` of silence before the frame.
    Silence,
    /// Large gap: full timeline reset.
    Reset,
}

/// Result of one evaluation.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Verdict {
    /// Action.
    pub action: PtsAction,
    /// The PTS to use (may differ from the input).
    pub corrected_pts: i64,
    /// Silence to insert (only for `Silence`).
    pub silence_ms: i32,
    /// Gap observed, in milliseconds (signed).
    pub gap_ms: i32,
}

/// Per-stream repair state.
#[derive(Debug, Clone)]
pub struct PtsRepair {
    last_pts: i64,
    last_duration: i64,
    tb_num: i32,
    tb_den: i32,
    small_gap_ms: i32,
    large_gap_ms: i32,
    last_gap_ms: i32,
    last_action_gap_ms: i32,
    consecutive_small_repairs: i32,
    relocking: bool,
    initialised: bool,
}

impl PtsRepair {
    /// `pts_repair_init`.
    pub fn new(small_gap_ms: i32, large_gap_ms: i32, tb_num: i32, tb_den: i32) -> Self {
        let _ = (small_gap_ms, large_gap_ms, tb_num, tb_den);
        todo!("W1-C")
    }

    /// `pts_repair_reset`.
    pub fn reset(&mut self) {
        let _ = (self.last_pts, self.last_duration, self.tb_num, self.tb_den, self.small_gap_ms, self.large_gap_ms,
                 self.last_gap_ms, self.last_action_gap_ms, self.consecutive_small_repairs, self.relocking, self.initialised);
        todo!("W1-C")
    }

    /// `pts_repair_evaluate`.
    pub fn evaluate(&mut self, pts: i64, duration: i64) -> Verdict {
        let _ = (pts, duration);
        todo!("W1-C")
    }

    /// Last known-good PTS and duration (stream time base), for extrapolating
    /// a frame that carries no timestamp.
    pub fn last(&self) -> Option<(i64, i64)> {
        todo!("W1-C")
    }
}
